controllers/operator_controller.go

```diff
@@ -63,7 +63,7 @@ type OperatorReconciler struct {
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.1/pkg/reconcile
 func (r *OperatorReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	l := log.FromContext(ctx).WithName("reconcile")
+	l := log.FromContext(ctx).WithName("operator-controller")
```

Comment:
Not for this PR (we can work on this in follow ups), but just confirming if I'm getting it right:

Since the operator controller hasn't been configured to receive events from catsrc controller, for now we have 2 isolated controllers, (ie):
- operator reconciler triggering with change in Operator object or BD, fetching entities from entity cache.
- Catsrc controller, working on populating the entity cache based on events it receives by watching catsrc on clusters.

So we have a possibility of operator controller reading a stale entity src cache, and giving us wrong results when a particular catsrc is modified?

Reply:
Yeah, at the moment it is possible that resolution can happen off a stale cache state. Most often this should only affect upgrades, which we don't handle yet. It is possible that a resolution can fail because the cache has not been populated yet. However, this should result in an error and a re-enqueue, iirc

Reply:
We'll definitely need to address this. But I think the resolver design will need to change if we want to do multi-stage resolution. It's likely that the resolver will need to be a separate process.

Reply:
> Most often this should only affect upgrades, which we don't handle yet. It is possible that a resolution can fail because the cache has not been populated yet. However, this should result in an error and a re-enqueue, iirc

Apart from upgrades, this scenario could also occur right?
- Catsrc controller updates the entity cache with say the entities for "prometheus" package.
- Operator Controller reconciles with the current state of entity cache successfully, installing the prometheus package on cluster.
- Catsrc is removed, catsrc controller reconciles and updates the entity cache.
- Nothing happens with operator controller since there is no re-trigger - unless either operator object is modified/ any events on BD trigger the reconciler/ the operator controller re-triggers after the default time interval.

The prometheus operator would still be available on cluster even after the respective package is removed from catsrc.

Which brings in my question on uninstall (sorry if it's a naive question, I'm not sure how OLM handles it currently - by subscriptions I guess?) -  what should trigger uninstall - the operator object being deleted or the change in packages in catsrc? Should the prometheus operator persist on cluster even after it is removed from catsrc even though the user has not explicitly mentioned an uninstall by deleting Operator object?

Reply:
Uninstall has been an achilles heel (besides other heels obviously) of OLM v0. TLDR is that the [uninstall process](https://olm.operatorframework.io/docs/tasks/uninstall-operator/) is cumbersome, and confusing. There have been many conversations that go
"I deleted the subscription, why did the operator not get uninstalled"
"Oh but the Subscription you created is a `intent` to subscribe to a particular channel for a package if you deleted the subscription we won't touch the workloads coz that's dangerous" 
"Okay that's not what I expected, when I created the Subscription the operator was installed, so when I delete the Subscription I expect the operator to be uninstalled"

That said, you bring up a good point. The install/uninstall flow for v1 warrants a larger discussion. 

At the face of it, it seems to me that we should only uninstall the operator if the Operator CR is deleted. If the CatalogSource is updated and the package is deleted from the CatalogSource, that should only trigger a notification in the Operator CR in the line of "The package that was installed from the CatalogSource foo was removed from foo, please get in touch with the CatalogSource owner to discuss the next steps if you want to receive updates for the operator you installed". 

We've also never had to think about scenarios where the package was removed from a CatalogSource, cause so far that hasn't happened. Maybe that helps us in deciding "okay we don't have to over optimize for that scenario just yet and we can go with some sane default scenario for now, as long as we keep scope for future enhancement for that feature, should that scenario occur".

Reply:
> > Most often this should only affect upgrades, which we don't handle yet. It is possible that a resolution can fail because the cache has not been populated yet. However, this should result in an error and a re-enqueue, iirc
> 
> Apart from upgrades, this scenario could also occur right?
> 
> * Catsrc controller updates the entity cache with say the entities for "prometheus" package.
> * Operator Controller reconciles with the current state of entity cache successfully, installing the prometheus package on cluster.
> * Catsrc is removed, catsrc controller reconciles and updates the entity cache.
> * Nothing happens with operator controller since there is no re-trigger - unless either operator object is modified/ any events on BD trigger the reconciler/ the operator controller re-triggers after the default time interval.
> 
> The prometheus operator would still be available on cluster even after the respective package is removed from catsrc.
> 
> Which brings in my question on uninstall (sorry if it's a naive question, I'm not sure how OLM handles it currently - by subscriptions I guess?) - what should trigger uninstall - the operator object being deleted or the change in packages in catsrc? Should the prometheus operator persist on cluster even after it is removed from catsrc even though the user has not explicitly mentioned an uninstall by deleting Operator object?

I would say that once a package is installed, even if it no longer exists anymore in a catsrc/cache, it should remain installed. IMHO the catalogsource is just a source of information. E.g. if we installed the prometheus operator directly from prometheus sources, and there's an outage, you wouldn't want prometheus to be uninstalled. Having said that, I think it would be good to include a condition on the operator that says whether the underlying source is still available.

Reply:
The way I think about uninstall, at least at the moment, is that if you delete the operator CR, everything goes! But it is a hairy landscape as Anik pointed out.